File: src/web/magmac/app/lang/PlantUMLLang.ts
import { Lists } from "../../../magmac/api/collect/list/Lists";
import { DivideRule } from "../../../magmac/app/compile/rule/DivideRule";
import { ExactRule } from "../../../magmac/app/compile/rule/ExactRule";
import { LocatingRule } from "../../../magmac/app/compile/rule/LocatingRule";
import { OrRule } from "../../../magmac/app/compile/rule/OrRule";
import { PrefixRule } from "../../../magmac/app/compile/rule/PrefixRule";
import { Rule } from "../../../magmac/app/compile/rule/Rule";
import { StringRule } from "../../../magmac/app/compile/rule/StringRule";
import { SuffixRule } from "../../../magmac/app/compile/rule/SuffixRule";
import { TypeRule } from "../../../magmac/app/compile/rule/TypeRule";
import { StatementFolder } from "../../../magmac/app/compile/rule/fold/StatementFolder";
export class PlantUMLLang {
	public static createRule() : Rule {
		return new DivideRule( "children", new StatementFolder( ), PlantUMLLang.createRootSegmentRule( ));
	}
	private static createRootSegmentRule() : SuffixRule {
		return new SuffixRule( new OrRule( Lists.of( CommonLang.createWhitespaceRule( ), new TypeRule( "start", new ExactRule( "@startuml\nskinparam linetype ortho")), new TypeRule( "end", new ExactRule( "@enduml")), PlantUMLLang.createStructureRule( "class"), PlantUMLLang.createStructureRule( "interface"), PlantUMLLang.createStructureRule( "enum"), new TypeRule( "inherits", LocatingRule.First( new StringRule( "parent"), " <|-- ", new StringRule( "child"))), new TypeRule( "dependency", LocatingRule.First( new StringRule( "child"), " --> ", new StringRule( "parent"))))), "\n");
	}
	private static createStructureRule( type : String) : Rule {
		return new TypeRule( type, new PrefixRule( type+" ", new StringRule( "name")));
	}
}

File: src/java/magmac/app/lang/PlantUMLLang.java
package magmac.app.lang;

import magmac.api.collect.list.Lists;
import magmac.app.compile.rule.NodeListRule;
import magmac.app.compile.rule.ExactRule;
import magmac.app.compile.rule.LocatingRule;
import magmac.app.compile.rule.OrRule;
import magmac.app.compile.rule.PrefixRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.StringRule;
import magmac.app.compile.rule.SuffixRule;
import magmac.app.compile.rule.TypeRule;
import magmac.app.compile.rule.fold.StatementFolder;
import magmac.app.lang.java.node.Whitespace;

public final class PlantUMLLang {
    public static Rule createRule() {
        return new NodeListRule("children", new StatementFolder(), PlantUMLLang.createRootSegmentRule());
    }

    private static SuffixRule createRootSegmentRule() {
        return new SuffixRule(new OrRule(Lists.of(
                Whitespace.createWhitespaceRule(),
                new TypeRule("start", new ExactRule("@startuml\nskinparam linetype ortho")),
                new TypeRule("end", new ExactRule("@enduml")),
                PlantUMLLang.createStructureRule("class"),
                PlantUMLLang.createStructureRule("interface"),
                PlantUMLLang.createStructureRule("enum"),
                new TypeRule("inherits", LocatingRule.First(new StringRule("parent"), " <|-- ", new StringRule("child"))),
                new TypeRule("dependency", LocatingRule.First(new StringRule("child"), " --> ", new StringRule("parent")))
        )), "\n");
    }

    private static Rule createStructureRule(String type) {
        return new TypeRule(type, new PrefixRule(type + " ", new StringRule("name")));
    }

}
